src/ds/rb_map.h

```diff
@@ -284,6 +303,353 @@ namespace rb
+          // - the right must also have a black height of 1 to maintain the
+          // height but it is not empty or we would have hit the above if
+          // statement
+          // - therefore the right node is red.
+          assert(r.left().empty());
+          assert(r.right().empty());
+          assert(r.rootColor() == R);
+          return std::make_pair(r.paint(B), false);
+        }
+        else if (right().empty())
+        {
+          // mirror of the above case
+          return std::make_pair(left().paint(B), false);
+        }
+        else
+        {
+          // both children are non-empty, swap this node's key and value with
+          // the successor and then delete the successor
+          auto successor = right().minimum();
+          auto right_without = right()._remove(successor.first);
+          auto newMap = Map(
+            rootColor(),
+            left(),
+            successor.first,
+            successor.second,
+            right_without.first);
+          if (right_without.second)
+          {
+            return newMap.fixDoubleBlackRight();
+          }
+          return std::make_pair(newMap, false);
+        }
+      }
+      else
+      {
+        // key not found in the tree
+        return std::make_pair(Map(*this), false);
+      }
+    }
+
+    // Return the minimum key in this map along with its value.
+    std::pair<K, V> minimum() const
+    {
+      assert(!empty());
+
+      if (left().empty())
+      {
+        return std::make_pair(rootKey(), rootValue());
+      }
+      else
+      {
+        return left().minimum();
+      }
+    }
+
+    // Check properties of the tree
+    void check() const
+    {
+      int totalBlackCount;
+      _check(0, totalBlackCount);
+    }
+
+    // Print an s-expression style representation of the tree's colors
+    void print() const
+    {
+      if (empty())
+      {
+        std::cout << "B";
+        return;
+      }
+      std::cout << "(";
+      std::cout << (rootColor() == B ? "B" : "R") << " ";
+      left().print();
+      right().print();
+      std::cout << ")";
+    }
+
+    void _check(int blackCount, int& totalBlackCount) const
+    {
+      if (empty())
+      {
+        totalBlackCount = blackCount + 1;
+        return;
+      }
+
+      if (rootColor() == R)
+      {
+        if (!left().empty() && left().rootColor() == R)
+        {
+          throw std::logic_error("rb::Map::check(): Double red node found");
+        }
+        if (!right().empty() && right().rootColor() == R)
+        {
+          throw std::logic_error("rb::Map::check(): Double red node found");
+        }
+      }
+      int leftBlackCount = 0;
+      int rightBlackCount = 0;
+      left()._check(blackCount + (rootColor() == B ? 1 : 0), leftBlackCount);
+      right()._check(blackCount + (rootColor() == B ? 1 : 0), rightBlackCount);
+
+      if (leftBlackCount != rightBlackCount)
+      {
+        print();
+        std::cout << std::endl;
```

Comment:
Remove?

Reply:
Would be neater to replace `print()` with `to_str()` (can retain streaming implementation by replacing `std::cout` with a `std::stringstream`, then returning `ss.str()`), then we can do the actual output (with inline newline) here. Not critical since its debug-only, but slightly cleaner.